runner/runner.go

```diff
@@ -4,61 +4,101 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
-	"log"
 	"net"
 	"net/http"
-	"strconv"
+	"os"
+	"os/signal"
 	"strings"
+	"syscall"
 	"time"
 
 	"github.com/gotify/server/v2/config"
 	"golang.org/x/crypto/acme/autocert"
+	"golang.org/x/sync/errgroup"
 )
 
 // Run starts the http server and if configured a https server.
 func Run(router http.Handler, conf *config.Configuration) {
-	httpHandler := router
-
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		done := make(chan os.Signal, 1)
+		signal.Notify(done, os.Interrupt, syscall.SIGTERM)
+		<-done
+		cancel()
+	}()
+	s := &http.Server{Handler: router}
+	g, gCtx := errgroup.WithContext(ctx)
 	if *conf.Server.SSL.Enabled {
-		if *conf.Server.SSL.RedirectToHTTPS {
-			httpHandler = redirectToHTTPS(strconv.Itoa(conf.Server.SSL.Port))
-		}
-
-		addr := fmt.Sprintf("%s:%d", conf.Server.SSL.ListenAddr, conf.Server.SSL.Port)
-		s := &http.Server{
-			Addr:    addr,
-			Handler: router,
-		}
-
 		if *conf.Server.SSL.LetsEncrypt.Enabled {
 			certManager := autocert.Manager{
 				Prompt:     func(tosURL string) bool { return *conf.Server.SSL.LetsEncrypt.AcceptTOS },
 				HostPolicy: autocert.HostWhitelist(conf.Server.SSL.LetsEncrypt.Hosts...),
 				Cache:      autocert.DirCache(conf.Server.SSL.LetsEncrypt.Cache),
 			}
-			httpHandler = certManager.HTTPHandler(httpHandler)
+			s.Handler = certManager.HTTPHandler(s.Handler)
 			s.TLSConfig = &tls.Config{GetCertificate: certManager.GetCertificate}
 		}
-		fmt.Println("Started Listening for TLS connection on " + addr)
-		go func() {
-			listener := startListening(addr, conf.Server.KeepAlivePeriodSeconds)
-			log.Fatal(s.ServeTLS(listener, conf.Server.SSL.CertFile, conf.Server.SSL.CertKey))
-		}()
+		g.Go(func() error {
+			return runTLS(s, conf)
+		})
+	}
+	g.Go(func() error {
+		return run(s, conf)
+	})
+	g.Go(func() error {
+		<-gCtx.Done()
```

Comment:
I found this difficult to understand with the multiple contexts, the extra g.Go for shutting down using gCtx.

I've tried a different approach with only a single channel that initiates the shutdown. This channel is used for the interrupts and the Serve errors. IMO it's easier to understand, what do you think? Here are the changes https://github.com/gotify/server/compare/unix-proposal?expand=1

I haven't fully tested it yet.

Reply:
Does this mean a socket file can be created but the net.Listen still returns an error?

Reply:
No, it would just return an error.

The issue is that the sockets won't be cleaned unless .Shutdown is called on the http server.

So, returning the error from TLS after the http is bound means there a chance it could not be cleaned up

Reply:
I'd expect the cleanup to be done on listener.Close which is already deferred. Could you describe your fix in more detail

> Easy fix is just a deferred function at the top once httplistener is binded so any return will run cleanup

Moving the listener creation to the goroutine should also fix the problem right, because the listener is always cleaned up in server.shutdown and then we only have on return from server.shutdown inside the Run function.
```go
go func() {
	httpListener, err := startListening("plain connection", conf.Server.ListenAddr, conf.Server.Port, conf.Server.KeepAlivePeriodSeconds)
	if err == nil {
		err = s.Serve(httpListener)
	}
	doShutdown(shutdown, err)
}()
```

Reply:
Maybe I am confusing myself because of the behaviors I was getting with `log.Fatal`, if defered .Closes are all what is needed then thats good. I will pull down your implementation now and run some tests. 👍🏻

Reply:
Works like a charm, so yeah the `log.Fatal`s caused confusion that the cleanup only happens on shutdown rather than close 👍🏻 

I ran my tests including acme stuff, works fine but I know you said you were going to test more on weekend. If you need me let me know. Thank you for spending time on time, I will now look at the trusted proxies. However, it seems a little more complicated with unix sockets but I figure something out.

Reply:
Yeah, it appears to be working, I've tested acme and all error combinations for listening on http and/or https.

Thanks for testing it too!

As a side note for the trusted proxies. I think gin-gonic does this by default, so it's possible to specify x-forwarded-for in the requests in this ip is then logged. But I'm not sure how this works with sockets.